File: src/main/java/proxy/Proxy.java
package proxy;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Random;

import pojo.Message;

enum Decision {
    NOMODIFY,
    IGNORE,
    MODIFY
}

public class Proxy {
    private static final int LISTENING_PORT = 50005;

    public static void main(String[] args) {
        ServerSocket serverSocket = null;
        try {
            serverSocket = new ServerSocket(LISTENING_PORT);
            System.out.println("Proxy listening on port " + LISTENING_PORT);
            while (true) {
                Socket socketFromClient = serverSocket.accept();
                ObjectInputStream sfcois = new ObjectInputStream(socketFromClient.getInputStream());
                Message messageFromClient = (Message) sfcois.readObject();

                System.out.println("Received from client: " + messageFromClient);

                InetSocketAddress serverAddress = messageFromClient.getHeader().getServerAddress();
                InetSocketAddress clientAddress = messageFromClient.getHeader().getClientAddress();

                Decision decisionToServer = decide();
                System.out.println("Let's " + decisionToServer.name() + " this message");
                if (decisionToServer == Decision.IGNORE) {
                    continue;
                }

                String messageToServerBody = decisionToServer == Decision.MODIFY ? new StringBuilder(messageFromClient.getBody()).reverse().toString() : messageFromClient.getBody();
                Message messageToServer = new Message(
                        new Message.Header(
                                clientAddress,
                                serverAddress
                        ),
                        messageToServerBody
                );

                System.out.println("Sending " + messageToServer);

                Socket socketToServer = new Socket(serverAddress.getHostName(), serverAddress.getPort());
                ObjectOutputStream stsoos = new ObjectOutputStream(socketToServer.getOutputStream());
                stsoos.writeObject(messageToServer);
                ObjectInputStream stsois = new ObjectInputStream(socketToServer.getInputStream());
                Message messageFromServer = (Message) stsois.readObject();
                stsoos.flush();
                stsoos.close();
                stsois.close();
                socketToServer.close();

                System.out.println("Received from server: \n"+messageFromServer);

                Decision decisionToClient = decide();
                System.out.println("Let's " + decisionToClient.name() + " this message");
                if (decisionToClient == Decision.IGNORE) {
                    continue;
                }

                String messageToClientBody = decisionToClient == Decision.MODIFY ? new StringBuilder(messageFromServer.getBody()).reverse().toString() : messageFromServer.getBody();
                Message messageToClient = new Message(
                        new Message.Header(
                                messageFromClient.getHeader().getServerAddress(),
                                messageFromClient.getHeader().getClientAddress()
                        ),
                        messageToClientBody
                );
                System.out.println("Sending " + messageToClient);

                ObjectOutputStream sfcoos = new ObjectOutputStream(socketFromClient.getOutputStream());
                sfcoos.writeObject(messageToClient);

                sfcoos.flush();
                sfcoos.close();
                socketFromClient.close();
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (serverSocket != null) {
                try {
                    serverSocket.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    private static Decision decide(){
        Random random = new Random();
        switch (random.nextInt(3)) {
            case 0: return Decision.NOMODIFY;
            case 1: return Decision.IGNORE;
            default: return Decision.MODIFY;
        }
    }
}

File: src/main/kotlin/proxy/Proxy.kt
package proxy

import pojo.Header
import pojo.Message
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.net.ServerSocket
import java.net.Socket
import kotlin.random.Random

const val LISTENING_PORT = 50005

enum class Decision {
    NOMODIFY,
    IGNORE,
    MODIFY
}

fun main(args: Array<String>) {
    ServerSocket(LISTENING_PORT).use {
        println("Proxy listening on port $LISTENING_PORT")
        while (true) {
            it.accept().use loop@ { socketFromClient ->
                val sfcois = ObjectInputStream(socketFromClient.getInputStream())
                val messageFromClient = sfcois.readObject() as Message

                println("Received from client: $messageFromClient")

                val serverAddress = messageFromClient.header.serverAddress
                val clientAddress = messageFromClient.header.clientAddress

                val decisionToServer = decide()
                println("Let's ${decisionToServer.name} this message")
                if (decisionToServer == Decision.IGNORE) {
                    return@loop
                }

                val messageToServerBody = if (decisionToServer == Decision.MODIFY) messageFromClient.body.reversed() else messageFromClient.body
                val messageToServer = Message(
                    Header(
                        clientAddress,
                        serverAddress
                    ),
                    messageToServerBody
                )

                println("Sending ${messageToServer}")

                val messageFromServer = Socket(serverAddress.hostName, serverAddress.port).use socketToServer@ {
                    val stsoos = ObjectOutputStream(it.getOutputStream())
                    stsoos.writeObject(messageToServer)

                    val stsois = ObjectInputStream(it.getInputStream())
                    return@socketToServer stsois.readObject() as Message
                }
                println("Received from server: \n$messageFromServer")

                val decisionToClient = decide()
                println("Lets ${decisionToClient.name} this message")
                if (decisionToClient == Decision.IGNORE) {
                    return@loop
                }

                val messageToClientBody = if (decisionToClient == Decision.MODIFY) messageFromServer.body.reversed() else messageFromServer.body
                val messageToClient = Message(
                    Header(
                        messageFromClient.header.serverAddress,
                        messageFromClient.header.clientAddress
                    ),
                    messageToClientBody
                )
                println("Sending $messageToClient")

                val sfcoos = ObjectOutputStream(socketFromClient.getOutputStream())
                sfcoos.writeObject(messageToClient)

            }
        }
    }
}

fun decide():Decision {
    val rI = Random.nextInt(3)
    return when (rI) {
        0 -> Decision.NOMODIFY
        1 -> Decision.IGNORE
        else -> Decision.MODIFY
    }
}